### PR TITLE
[cli] Uploads: add progress bar

### PR DESCRIPTION
**Pull request by julien-c**
see https://github.com/huggingface/transformers/pull/2044#discussion_r354057827 for context

There might be a more pythonic way (to do a "simple" method overriding) but I couldn't find it.

### CODE REVIEW

File: transformers/hf_api.py [hunk 1]
```diff
@@ -148,6 +152,27 @@ def list_objs(self, token):
 
 
 
+class ProgressFileReader:
```
Comment (rlouf) [inline transformers/hf_api.py]: Definitely non-standard, but definitely clever :)

I don't think many people would understand this behavior, as you have to know values are passed by reference in Python, and be familiar with `io.BufferedReader` 's API. I would document the `ProgressFileReader` class to explain why we are doing this and why it works.

A more pythonic API would be to promote `ProgressFileReader` to a [context manager](https://docs.python.org/3/library/contextlib.html)  and open the file in this context manager so the uploading code would look like:

```python
with stream_with_progressbar(filepath, "rb") as f:
       r = requests.put(urls.write, data=f, headers={
              "content-type": urls.type,
       })
       r.raise_for_status()
```

Also has the merit of being self-documenting. The closing logic is handled inside the context manager.

**Review by thomwolf — APPROVED** [review]
I like it, I think it's fine.

Might even be reused in other locations of the library and examples later so maybe put it in `file_utils.py` and add it to `__init__`?

File: transformers/hf_api.py [hunk 2]
```diff
@@ -129,10 +130,13 @@ def presign_and_upload(self, token, filename, filepath):
         # Even though we presign with the correct content-type,
         # the client still has to specify it when uploading the file.
         with open(filepath, "rb") as f:
+            pf = ProgressFileReader(f)
```
Comment (thomwolf) [inline transformers/hf_api.py]: Maybe call it `TqdmProgressFileReader` if it's included in `file_utils`?

File: transformers/hf_api.py (hunk 1 above)
Comment (julien-c) [reply transformers/hf_api.py]: I could do a context manager, but I can't encapsulate `f` completely like in your code sample, (passing `f` to `requests.put`) as I don't really know which methods of the io.BufferedReader `requests` (and its underlying libraries) is going to call.

The cleanest way would be to subclass `io.BufferedReader` and instantiate an instance of the child class with the parent instance, but I don't know how to do that in Python.

File: transformers/hf_api.py (hunk 2 above)
Comment (julien-c) [reply transformers/hf_api.py]: renamed it, let's move it to file_utils if/when we need it somewhere else.

**Pull request merged**